ext/gmp/tests/gmp_cryptography_ffc.phpt

```diff
@@ -0,0 +1,108 @@
+--TEST--
+Examples of the usage of gmp for finite field cryptography.
+--DESCRIPTION--
+This executes basic operations (addition, multiplication, inverse, exponentiation) as the "base operations".
+Then, it performs a primality check, and finally diffie-hellman as the "application".
+All operations are done in the 4096-bit MODP Group from RFC 3526: https://www.ietf.org/rfc/rfc3526.txt
+
+Omitted are calculations on elliptic curves, which are also common, because of the complexity of these algorithms.
+Elliptic curves generally operate on smaller values, so their use-case is somewhat covered here,
+but curve calculations may need additional operations not used here.
+
+Further, omitted is explicit demonstration of (public-key) encryption, commitments, zero-knowledge proofs or similar common applications.
+However, the operation used in the diffie-hellman is at the core of all these other applications, hence these use-cases are implicitly covered, too.
+
+$a, $b, and $c generated with
+$random = gmp_random_range(0, $prime);
+$randomHex = strtoupper(gmp_strval($random, 16));
+echo chunk_split(chunk_split($randomHex, 8, " "), 54);
+--EXTENSIONS--
+gmp
+--FILE--
+<?php
+
+$prime = gmp_init('
+FFFFFFFF FFFFFFFF C90FDAA2 2168C234 C4C6628B 80DC1CD1
+29024E08 8A67CC74 020BBEA6 3B139B22 514A0879 8E3404DD
+EF9519B3 CD3A431B 302B0A6D F25F1437 4FE1356D 6D51C245
+E485B576 625E7EC6 F44C42E9 A637ED6B 0BFF5CB6 F406B7ED
+EE386BFB 5A899FA5 AE9F2411 7C4B1FE6 49286651 ECE45B3D
+C2007CB8 A163BF05 98DA4836 1C55D39A 69163FA8 FD24CF5F
+83655D23 DCA3AD96 1C62F356 208552BB 9ED52907 7096966D
+670C354E 4ABC9804 F1746C08 CA237327 FFFFFFFF FFFFFFFF', 16);
```

Comment:
Are spaces actually accepted within hex strings?

Reply:
Yes:

```
$a = gmp_init('FFFFFFFF FFFFFFFF', 16);
$b = gmp_init('FFFFFFFFFFFFFFFF', 16);

echo var_dump(gmp_cmp($a, $b) === 0);
```

will output `bool(true)`.

I am writing it this way here as it corresponds to how it is written in the RFC.

Reply:
I just did not know this was possible, probably should open a doc-en issue about this to improve the documentation.